### ISSUE TITLE
Better handling of large combinatorial experiments

### ISSUE THREAD

**Issue by andrrizzi** [choderalab/yank#852]
Babysitting very large combinatorial simulations can be painful right now. If only some of the calculations fail for some reason (e.g., cluster problems), we can only do one of the following:
1. Restart every combinatorial experiment. This is currently slow because we need to load the full `ReplicaExchange` object from disk to determine whether a specific phase is completed or not.
2. For job arrays, we have to manually dig in the logs to find the jobid of the failed experiments and run only those. This is obviously not ideal.

I'd like to make two changes
1. Add a class method in `ReplicaExchange` (and relative instance method in `AlchemicalPhase`) to check the status of the calculations without loading the full `ReplicaExchange` object. Something like:
```python
class ReplicaExchangeStatus(typing.NamedTuple):
    """Status of the replica-exchange calculation"""
    number_of_iterations: int
    is_completed: bool

@classmethod
def get_status(cls, storage):
    """Read information about the status of the calculation from the storage file.

    Parameters
    ----------
    storage : str or Reporter
        The path to the storage file or the reporter object.

    Returns
    -------
    status : ReplicaExchangeStatus
        The status of the calculation.
    """
```
2. Resuscitate the `yank status` CLI argument (currently broken) to optionally take a YAML script as an input instead of the directory of a single experiment, and print status information for all experiments. The output would look like
```
yank status --yaml=yank.yaml
Total number of experiments: 155 (25 completed, 75 running, 55 pending)
```
or with verbose output
```
yank status --yaml=yank.yaml -v
experiment1:
    phase1: number_of_iterations=10000, status=completed
    phase2: number_of_iterations=10000, status=completed
...
experiment55:
    phase1: number_of_iterations=10000, status=completed
    phase2: number_of_iterations=2458, status=running
...
experiment124:
    phase1: status=pending
    phase2: status=pending
...
Total number of experiments: 155 (25 completed, 75 running, 55 pending)
```
I'd also add the argument `--status` to print only info about completed/running/pending experiments, and the argument `--njobs` to associate a `jobid` to the experiments, which can be helpful for array jobs.

I think this could save us a lot of troubles in the next couple of months. Objections? Thoughts?

**Comment by jchodera** [comment]
I'm all in favor of increasing usability!

How would you imagine users would "fix" the broken combinatorial simulations and revive them? Is this just to recover from NaNs (which it seems like we should just find a better way to do automatically), or do they actually have to fix some input files and re-run?

**Comment by jchodera** [comment]
If it's just really to recover from cluster issues, then shouldn't we have built-in fault tolerance to try again?

**Comment by andrrizzi** [comment]
To give you a concrete example, what's currently giving me a headache is that I get `Exception: No compatible CUDA device is available` on some nodes (I'm in touch with Sveta to try to figure this out). I don't see a way to include in YANK something that gets the job rescheduled on a different host, but it'd be very easy with these changes in to just add `select[hname!=incriminatedhost]` and restart.

**Comment by jchodera** [comment]
> To give you a concrete example, what's currently giving me a headache is that I get Exception: No compatible CUDA device is available on some nodes (I'm in touch with Sveta to try to figure this out).

Oof. This is a really serious error because it means some nodes will just eat all the jobs in the queue and cause them to fail. Can you CC me on that HPC ticket? These nodes should really *remove* themselves from the queue when that happens lest they eat an entire stack of queued jobs.

> I don't see a way to include in YANK something that gets the job rescheduled on a different host, but it'd be very easy with these changes in to just add select[hname!=incriminatedhost] and restart.

I see: You're just looking to reduce the time required for restarts.

If this is going to be a persistent problem, we *could* add a "check_gpus" command to the [MPI framework](https://github.com/choderalab/yank/blob/master/Yank/mpi.py) you developed which would return a new MPI communicator restricted to *just* the MPI processes with working GPUs. That call could simply be inserted at the very beginning of YANK, and could do a quick GPU test of some reference system to ensure the process has a valid GPU attached.

**Comment by andrrizzi** [comment]
> If this is going to be a persistent problem, we could add a "check_gpus" command to the MPI framework

I like it! In the meantime, I'll cc you on the HPC ticket and work on adding these changes.

**Comment by andrrizzi** [comment]
This was implemented in #853.
